File: code/Python-scripts-working/05_gsea/01_gsea.py
"""STEP5 : Hallmark GSEA + KEGG ssGSEA (R 01_gsea.R 이식).

(1) Hallmark GSEA — GSE142025 DEG 3비교의 logFC 랭킹, gseapy.prerank (gmt 4-1).
(2) KEGG ssGSEA — GSE96804 발현, gseapy.ssgsea (gmt 4-2).
FN1(EMT/산화)·ALDH2(산화/대사) 관련 경로 유의성 확인.
라이브러리 매핑: clusterProfiler::GSEA → gseapy.prerank, GSVA ssGSEA → gseapy.ssgsea.
출력: RES_DIR/step5_gsea/
"""
import sys, pathlib
sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1]))
import config
import pandas as pd

OUT = config.RES_DIR / "step5_gsea"; OUT.mkdir(parents=True, exist_ok=True)
S1  = config.RES_DIR / "step1_deg"                    # 원본과 동일 diff_ 입력 위치
HALLMARK_GMT = config.DATA_ROOT / "4-1. h.all.v2026.1.Hs.symbols.gmt"
KEGG_GMT     = config.DATA_ROOT / "4-2. c2.cp.kegg_legacy.v2026.1.Hs.symbols.gmt"
FOCUS = "EPITHELIAL_MESENCHYMAL|REACTIVE_OXYGEN|OXIDATIVE|INFLAMMATORY|TGF_BETA|APOPTOSIS"
KFOCUS = "ECM_RECEPTOR|CITRATE_CYCLE|OXIDATIVE_PHOSPHORYLATION|FOCAL_ADHESION|TRYPTOPHAN"
TAGS = ["Late_vs_Early", "Late_vs_Control", "Early_vs_Control"]


def _prerank(rnk, gmt):
    import gseapy as gp
    return gp.prerank(rnk=rnk, gene_sets=str(gmt), min_size=5, max_size=1000,
                      permutation_num=1000, seed=123, outdir=None, no_plot=True).res2d


def _run_gsea(gmt, prefix, focus):
    """원본 GSEA.R/hallmark.gsea.R 와 동일: diff_(유의 DEG) logFC 랭킹으로 GSEA."""
    try:
        import gseapy as gp  # noqa
    except Exception:
        print(f"[{prefix}] TODO: gseapy 미설치 → 'pip install gseapy'."); return
    for tag in TAGS:
        f = S1 / f"DEG_diff_{tag}.txt"                # 원본과 동일 diff_ 입력
        if not f.exists():
            print(f"[{prefix} {tag}] diff DEG 없음 → STEP1 02_deg 먼저: {f}"); continue
        deg = pd.read_csv(f, sep="\t")
        rnk = deg[["id", "logFC"]].dropna().drop_duplicates("id").sort_values("logFC", ascending=False)
        tab = _prerank(rnk, gmt)
        tab.to_csv(OUT / f"{prefix}.GSEA.{tag}.txt", sep="\t", index=False)
        term_col = "Term" if "Term" in tab.columns else tab.columns[0]
        foc = tab[tab[term_col].str.contains(focus, regex=True, na=False)]
        print(f"[{prefix} {tag}] ranked {len(rnk)} | 총 {len(tab)} | 관련 {len(foc)}행")


def run_hallmark():
    _run_gsea(HALLMARK_GMT, "Hallmark", FOCUS)


def run_kegg_gsea():
    """원본 GSEA.R 대응(신규): 동일 diff_ DEG 를 KEGG gmt 로 GSEA."""
    _run_gsea(KEGG_GMT, "KEGG", KFOCUS)


def run_kegg_ssgsea():
    try:
        import gseapy as gp
    except Exception:
        print("[KEGG ssGSEA] TODO: gseapy 미설치."); return
    expr_f = config.OUT_DIR / "GSE96804.labeled.txt"
    if not expr_f.exists():
        print("[KEGG ssGSEA] GSE96804 매트릭스 없음 → R STEP2 필요."); return
    expr = pd.read_csv(expr_f, sep="\t", index_col=0)
    ss = gp.ssgsea(data=expr, gene_sets=str(KEGG_GMT), outdir=None, sample_norm_method="rank", no_plot=True)
    scores = ss.res2d
    scores.to_csv(OUT / "KEGG.ssGSEA.scores.txt", sep="\t", index=False)
    print("[KEGG ssGSEA] 저장. FN1/ALDH2 ↔ 경로 상관 계산 중...")

    # --- FN1/ALDH2 ↔ 경로 Spearman 상관 (R 대조: FN1↔ECM 0.919, ALDH2↔TCA 0.921) ---
    from scipy.stats import spearmanr
    mat = scores.pivot(index="Term", columns="Name", values="NES")   # 경로 × 샘플
    samples = [s for s in mat.columns if s in expr.columns]; mat = mat[samples]
    rows = []
    for g in ["FN1", "ALDH2"]:
        if g not in expr.index:
            continue
        gv = expr.loc[g, samples].astype(float).values
        cs = pd.Series({p: spearmanr(gv, mat.loc[p, samples].astype(float).values).correlation
                        for p in mat.index}).sort_values(ascending=False)
        for p, r in cs.head(5).items():
            rows.append({"gene": g, "pathway": p, "rho": round(float(r), 3), "kind": "top5"})
        # R 대조용 특정 경로 명시(top5 밖이어도 기록)
        for p in ["KEGG_ECM_RECEPTOR_INTERACTION", "KEGG_CITRATE_CYCLE_TCA_CYCLE"]:
            if p in cs.index:
                rows.append({"gene": g, "pathway": p, "rho": round(float(cs[p]), 3), "kind": "R대조"})
        print(f"  [{g}] top: {cs.index[0]} rho={cs.iloc[0]:.3f}")
    pd.DataFrame(rows).to_csv(OUT / "FN1_ALDH2_pathway_corr.csv", index=False)


def main():
    run_hallmark()
    run_kegg_gsea()       # 원본 GSEA.R 대응(신규)
    run_kegg_ssgsea()     # 논문 Fig2H,I


if __name__ == "__main__":
    main()

File: code/R-scripts-working/05_gsea/01_gsea.R
# 01_gsea.R -----------------------------------------------------------------
# STEP 5 : GSEA(Hallmark+KEGG) + ssGSEA.  원본 GSEA.R / hallmark.gsea.R 1:1 대조.
#   (1) Hallmark GSEA — 원본 hallmark.gsea.R 와 동일: GSE142025 DEG 3비교의 **diff_(유의 DEG)**
#       logFC 랭킹 → GSEA(TERM2GENE, pvalueCutoff=1). 제공 gmt(4-1) 사용(원본 msigdbr 대체, 오프라인).
#   (2) KEGG GSEA — 원본 GSEA.R 와 동일: 동일 diff_ DEG 를 KEGG gmt(4-2)로 GSEA. (기존 파이프라인 누락분 추가)
#   (3) KEGG ssGSEA — 논문 Fig 2H,I(단계별 경로점수) 방법: GSE96804 발현에 ssGSEA →
#       DKD vs Control Wilcoxon + FN1/ALDH2 발현 상관(멀티오믹스 연결). 원본 GSEA.R(=GSEA) 와는 다른 분석.
#   출력: results/step5_gsea/. 원본 데이터·스크립트 무수정.
# ---------------------------------------------------------------------------

suppressMessages({
  library(clusterProfiler); library(enrichplot); library(GSVA)
  library(ggplot2); library(limma)
})
library(here)
source(here::here("config.R"))
OUT <- file.path(RES_DIR, "step5_gsea"); dir.create(OUT, showWarnings = FALSE, recursive = TRUE)
set.seed(123)

HALLMARK_GMT <- file.path(DATA_ROOT, "4-1. h.all.v2026.1.Hs.symbols.gmt")
KEGG_GMT     <- file.path(DATA_ROOT, "4-2. c2.cp.kegg_legacy.v2026.1.Hs.symbols.gmt")

## read.gmt 컬럼명 버전차 방어 (term/gene 로 표준화)
read_gmt_std <- function(path) {
  g <- read.gmt(path)
  names(g)[1:2] <- c("term", "gene")
  g
}

## ================= (1) Hallmark GSEA =================
hm <- read_gmt_std(HALLMARK_GMT)   # term2gene: (term, gene)
message("[Hallmark] gene sets: ", length(unique(hm$term)))

# 원본 GSEA.R/hallmark.gsea.R 와 동일: diff_(유의 DEG) 의 logFC 랭킹으로 GSEA.
read_diff_lfc <- function(degFile) {
  d <- read.table(file.path(RES_DIR, "step1_deg", degFile), header = TRUE, sep = "\t", row.names = 1, check.names = FALSE)
  lfc <- d$logFC; names(lfc) <- rownames(d)
  sort(lfc[!is.na(lfc) & !duplicated(names(lfc))], decreasing = TRUE)
}
run_gsea <- function(degFile, tag, term2gene, prefix) {
  lfc <- read_diff_lfc(degFile)
  kk <- GSEA(lfc, TERM2GENE = term2gene, pvalueCutoff = 1, seed = TRUE, verbose = FALSE)
  tab <- as.data.frame(kk)
  write.table(tab, file.path(OUT, paste0(prefix, ".GSEA.", tag, ".txt")), sep = "\t", quote = FALSE, row.names = FALSE)
  message("[", prefix, " ", tag, "] ranked ", length(lfc), " genes | total ", nrow(tab),
          " | sig(p.adj<0.05) ", sum(tab$p.adjust < 0.05))
  tab
}
# 원본과 동일하게 diff_ (유의 DEG) 입력
cmp <- list(Early_vs_Control = "DEG_diff_Early_vs_Control.txt",
            Late_vs_Control  = "DEG_diff_Late_vs_Control.txt",
            Late_vs_Early    = "DEG_diff_Late_vs_Early.txt")
hres <- Map(function(f, n) run_gsea(f, n, hm, "Hallmark"), cmp, names(cmp))

# FN1/ALDH2 관련 Hallmark 경로(EMT/ROS/염증 등) 요약
focus_pat <- "EPITHELIAL_MESENCHYMAL|REACTIVE_OXYGEN|OXIDATIVE|INFLAMMATORY|TGF_BETA|APOPTOSIS"
cat("\n== Hallmark FN1/ALDH2 관련 경로 (NES/p.adjust) ==\n")
focusTab <- do.call(rbind, lapply(names(hres), function(n) {
  t <- hres[[n]]; t <- t[grepl(focus_pat, t$ID), c("ID","NES","pvalue","p.adjust")]
  if (nrow(t)) cbind(Comparison = n, t) else NULL
}))
print(focusTab, row.names = FALSE)
write.csv(focusTab, file.path(OUT, "Hallmark_focus_pathways.csv"), row.names = FALSE)

# 논문 대조 CSV (Hallmark EMT/OXPHOS NES) — 논문 Fig2: EMT Late_vs_Ctrl +2.52 / Late_vs_Early +2.59, OXPHOS -1.62 / -2.02
getNES <- function(comp, pat) { t <- hres[[comp]]; v <- t$NES[grepl(pat, t$ID)]; if (length(v)) v[1] else NA }
hall_cmp <- data.frame(
  pathway = c("EMT", "EMT", "OXIDATIVE_PHOSPHORYLATION", "OXIDATIVE_PHOSPHORYLATION"),
  comparison = c("Late_vs_Control", "Late_vs_Early", "Late_vs_Control", "Late_vs_Early"),
  paper_NES = c(2.52, 2.59, -1.62, -2.02),
  ours_NES = c(getNES("Late_vs_Control", "EPITHELIAL_MESENCHYMAL"),
               getNES("Late_vs_Early",   "EPITHELIAL_MESENCHYMAL"),
               getNES("Late_vs_Control", "OXIDATIVE_PHOSPHORYLATION"),
               getNES("Late_vs_Early",   "OXIDATIVE_PHOSPHORYLATION")))
hall_cmp$ours_NES <- round(hall_cmp$ours_NES, 2)
write.csv(hall_cmp, file.path(OUT, "compare_paper_vs_ours.Hallmark_NES.csv"), row.names = FALSE)
cat("\n== Hallmark NES 논문 대조 ==\n"); print(hall_cmp, row.names = FALSE)

## ================= (2) KEGG GSEA (원본 GSEA.R) =================
kegg <- read_gmt_std(KEGG_GMT)
message("\n[KEGG GSEA] gene sets: ", length(unique(kegg$term)))
kres <- Map(function(f, n) run_gsea(f, n, kegg, "KEGG"), cmp, names(cmp))
# FN1(ECM)·ALDH2(TCA/대사) 관련 KEGG 경로 요약
kfocus_pat <- "ECM_RECEPTOR|CITRATE_CYCLE|TCA|OXIDATIVE_PHOSPHORYLATION|FOCAL_ADHESION|TRYPTOPHAN|ARGININE"
kfocusTab <- do.call(rbind, lapply(names(kres), function(n) {
  t <- kres[[n]]; t <- t[grepl(kfocus_pat, t$ID), c("ID","NES","pvalue","p.adjust")]
  if (nrow(t)) cbind(Comparison = n, t) else NULL
}))
if (!is.null(kfocusTab)) { cat("\n== KEGG GSEA FN1/ALDH2 관련 경로 ==\n"); print(kfocusTab, row.names = FALSE)
  write.csv(kfocusTab, file.path(OUT, "KEGG.GSEA_focus_pathways.csv"), row.names = FALSE) }

## ================= (3) KEGG ssGSEA (논문 Fig 2H,I) =================
keggList <- split(kegg$gene, kegg$term)
message("\n[KEGG ssGSEA] gene sets: ", length(keggList))

expr <- as.matrix(read.table(file.path(OUT_DIR, "GSE96804.labeled.txt"),
                             header = TRUE, sep = "\t", check.names = FALSE, row.names = 1))
grp <- ifelse(grepl("_Control$", colnames(expr)), "Control", "DKD")

par_ss <- ssgseaParam(expr, keggList)
ss <- gsva(par_ss)                       # 경로 x 샘플 점수
write.table(rbind(pathway = colnames(t(ss)), t(ss)), file.path(OUT, "KEGG.ssGSEA.scores.txt"),
            sep = "\t", quote = FALSE, col.names = FALSE)
message("[KEGG ssGSEA] 경로 x 샘플 = ", nrow(ss), " x ", ncol(ss))

# DKD vs Control Wilcoxon per pathway
wilcox_tab <- data.frame(pathway = rownames(ss),
  p = apply(ss, 1, function(v) wilcox.test(v ~ grp)$p.value),
  meanDKD = rowMeans(ss[, grp == "DKD", drop = FALSE]),
  meanCtrl = rowMeans(ss[, grp == "Control", drop = FALSE]))
wilcox_tab$padj <- p.adjust(wilcox_tab$p, "BH")
wilcox_tab$deltaDKD <- wilcox_tab$meanDKD - wilcox_tab$meanCtrl
wilcox_tab <- wilcox_tab[order(wilcox_tab$padj), ]
write.csv(wilcox_tab, file.path(OUT, "KEGG.ssGSEA.DKDvsControl.csv"), row.names = FALSE)
cat("\n== KEGG ssGSEA DKD vs Control (top 12 by padj) ==\n")
print(head(wilcox_tab[, c("pathway","deltaDKD","padj")], 12), row.names = FALSE)

# FN1/ALDH2 발현과 경로점수 상관
cor_focus <- function(gene) {
  if (!(gene %in% rownames(expr))) return(NULL)
  g <- as.numeric(expr[gene, ])
  cc <- apply(ss, 1, function(v) cor(v, g, method = "spearman"))
  data.frame(gene = gene, pathway = names(cc), rho = as.numeric(cc))
}
corTab <- rbind(cor_focus("FN1"), cor_focus("ALDH2"))
corTab <- corTab[order(-abs(corTab$rho)), ]
write.csv(corTab, file.path(OUT, "KEGG.ssGSEA.FN1_ALDH2_corr.csv"), row.names = FALSE)
cat("\n== FN1/ALDH2 vs KEGG ssGSEA 상관 상위 (|rho|) ==\n")
print(head(corTab, 12), row.names = FALSE)

message("\n[STEP5] 완료 -> ", OUT)
